contributors/design-proposals/architecture/scope.md

```diff
@@ -0,0 +1,342 @@
+  example, things without which users otherwise were
+  reverse-engineering Kubernetes to figure out, and/or copying code
+  out of Kubernetes itself to make work.
+
+* Is it functionality that users expect, such as because other
+  container platforms and/or service discovery and routing mechanisms
+  provide it? If a capability that relates to Kubernetes's fundamental
+  purpose were to become table stakes in the industry, Kubernetes
+  would need to support it in order to stay relevant. (Whether it
+  would need to be addressed by the core project would depend on the
+  other criteria.)
+
+* Is there sufficiently broad user demand and/or sufficient expected
+  user benefit for the functionality?
+
+* Is there an adequate mechanism to discover, deploy, express a
+  dependency on, and upgrade the functionality if implemented using an
+  extension mechanism? Are there consistent notions of releases, maturity,
+  quality, version skew, conformance, etc. for extensions?
+
+* Is it needed as a reference implementation exercising extension
+  points or other APIs?
+
+* Is the functionality sufficiently general-purpose?
+
+* Is it an area where we want to provide an opinionated solution
+  and/or where fragmentation would be problematic for users, or are
+  there many reasonable alternative approaches and solutions to the
+  problem?
+
+* Is it an area where we want to foster exploration and innovation in
+  the ecosystem?
+
+* Has the ecosystem produced adequate solutions on its own? For
+  instance, have ecosystem projects taken on requirements of the
+  Kubernetes project, if needed? Example: etcd3 added a number of features
+  and other improvements to benefit Kubernetes, so the project didn't
+  need to launch a separate storage effort.
+
+* Is there an acceptable home for the recommended ecosystem solution(s)?
+  Example: the [CNCF Sandbox](https://github.com/cncf/toc/blob/master/process/sandbox.md) is one possible home
+
+* Has the functionality been provided by the project/release/component
+  historically?
+
+## Technical scope details and rationale
+
+### Containerized workload execution and management
+
+Including:
+* common general categories of workloads, such as stateless, stateful, batch, and cluster services
+* provisioning, allocation, accessing, and managing compute, storage, and network resources on behalf of the workloads, and enforcement of security policies on those resources
+* workload prioritization, capacity assessment, placement, and relocation (aka scheduling)
+* graceful workload eviction
+* local container image caching
+* configuration and secret distribution
+* manual and automatic horizontal and vertical scaling
+* deployment, progressive (aka rolling) upgrades, and downgrades
+* self-healing
+* exposing container logs, status, health, and resource usage metrics for collection
+
+### Service discovery, load balancing, and routing
+
+Including:
+* endpoint tracking and discovery, including pod and non-pod endpoints
```

Comment:
We have affordances for external services today, but I wonder whether that was a mistake.  I am not convinced we want to carry that to logical conclusion, as opposed to integrating into higher-level SD systems, including things like Consul but also things like Istio.

I guess, since we have it now it's hard to say it's out of scope, but I am not 100% sure we want to carry it forward.

Reply:
What I think was a mistake is creating the expectation that Endpoints would be durable -- it should not be the source of record. If we nuked all Endpoints, we should expect controllers to repopulate them.

It's worth thinking about what interfaces we'd want if we intended to integrate with Consul and Istio, but we do need an interface that's accessible by more than just proxies. Monitoring and health-checking systems are obvious examples of other consumers of the information. 

Otherwise the ecosystem will fragment and/or lots of things will then track pods directly, and label selectors would need to be propagated by users and implemented in many clients. (We still need an API-based mechanism to convert the schematized form of label selectors to the query param form: https://github.com/kubernetes/kubernetes/issues/3676.)

Reply:
Also ref https://github.com/kubernetes/kubernetes/issues/18610

Reply:
Endpoints slide of contributor summit deck: https://docs.google.com/presentation/d/16c8i5ptRcsubBE_uRymGTklgOcvmAeAuRxSsSIWKGeM/edit#slide=id.g4741dd49cd_1_0